modules/bastion/main.tf

```
@@ -0,0 +1,139 @@
+locals {
+  enabled         = module.this.enabled
+  route53_enabled = local.enabled && var.associate_public_ip_address && var.custom_bastion_hostname != null && var.vanity_domain != null
+
+  vpc_id                 = module.vpc.outputs.vpc_id
+  vpc_private_subnet_ids = module.vpc.outputs.private_subnet_ids
+  vpc_public_subnet_ids  = module.vpc.outputs.public_subnet_ids
+  vpc_subnet_ids         = var.associate_public_ip_address ? local.vpc_public_subnet_ids : local.vpc_private_subnet_ids
+
+  bastion_subnet = slice(local.vpc_subnet_ids, 0, 1)
+  bastion_az     = module.vpc.outputs.availability_zones[0]
+
+  userdata_template = "${path.module}/templates/user-data.sh"
+
+  cloudwatch_agent_config = <<-END
+    #cloud-config
+    ${jsonencode({
+  write_files = [
+    {
+      path        = "/tmp/container.sh"
+      permissions = "0755"
+      owner       = "root:root"
+      encoding    = "b64"
+      content = base64encode(templatefile("${path.module}/templates/container.sh"), {
+        region           = var.region
+        image_repository = var.image_repository
+        image_container  = var.image_container
+      })
+    },
+  ]
+})}
+  END
+}
+
+data "aws_route53_zone" "vanity" {
+  count = local.route53_enabled ? 1 : 0
+  name  = var.vanity_domain
+}
+
+resource "aws_route53_record" "default" {
+  count   = local.route53_enabled ? 1 : 0
+  zone_id = join("", data.aws_route53_zone.vanity.*.zone_id)
+  name    = "${var.custom_bastion_hostname}.${var.vanity_domain}"
+  type    = "A"
+  ttl     = "300"
+  records = aws_eip.static.*.public_ip
+}
+
+module "aws_key_pair" {
+  source              = "cloudposse/key-pair/aws"
+  version             = "0.18.0"
+  attributes          = ["ssh", "key"]
+  ssh_public_key_path = var.ssh_key_path
+  generate_ssh_key    = var.generate_ssh_key
+
+  context = module.this.context
+}
+
+data "cloudinit_config" "config" {
+  count         = local.enabled ? 1 : 0
+  gzip          = false
+  base64_encode = true
+
+  part {
+    content_type = "text/cloud-config"
+    filename     = "cloud-config.yaml"
+    content      = local.cloudwatch_agent_config
+  }
+
+  part {
+    content_type = "text/x-shellscript"
+    filename     = "user-data.sh"
+    content      = templatefile(local.userdata_template, {
+      ssh_pub_keys     = var.ssh_pub_keys
+    })
+  }
+}
+
+resource "aws_eip" "static" {
+  count = local.enabled ? 1 : 0
+  vpc   = true
+}
+
+module "ec2_bastion" {
+  source  = "cloudposse/ec2-bastion-server/aws"
+  version = "0.28.0"
+
+  instance_type = var.instance_type
+  ## Use only one availability zone to be sure volume will be in the same zone
+  subnets                       = local.bastion_subnet
+  vpc_id                        = local.vpc_id
+  root_block_device_volume_size = var.root_block_device_volume_size
+  associate_public_ip_address   = true
```

```suggestion
  associate_public_ip_address   = var.associate_public_ip_address
```

`var.associate_public_ip_address` is ignored, despite even being passed in the example snippet.